### ISSUE TITLE
[CI failure] DateHistogramTests, "The datetime zone id 'CET' is not recognised"

### ISSUE THREAD

**Issue by cbuescher** [elastic/elasticsearch#14524]
Cannot reproduce this so far, but there are test failures in
* org.elasticsearch.messy.tests.DateHistogramTests.testDSTBoundaryIssue9491
* org.elasticsearch.messy.tests.DateHistogramTests.testIssue8209

http://build-us-00.elastic.co/job/es_core_2x_small/1726

```
org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.raiseEarlyFailure(TransportSearchTypeAction.java:316)
	... 10 more
Caused by: java.lang.IllegalArgumentException: The datetime zone id 'Asia/Jerusalem' is not recognised
	at org.joda.time.DateTimeZone.forID(DateTimeZone.java:229)
	at org.elasticsearch.search.aggregations.support.ValuesSourceParser.token(ValuesSourceParser.java:119)
	at org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramParser.parse(DateHistogramParser.java:97)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:198)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:103)
	at org.elasticsearch.search.aggregations.AggregationParseElement.parse(AggregationParseElement.java:60)
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:835)
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:655)
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:621)
	at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:266)
	at 
```

and

```
...
Caused by: java.lang.IllegalArgumentException: The datetime zone id 'CET' is not recognised
	at org.joda.time.DateTimeZone.forID(DateTimeZone.java:229)
	at org.elasticsearch.search.aggregations.support.ValuesSourceParser.token(ValuesSourceParser.java:119)
	at org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramParser.parse(DateHistogramParser.java:97)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:198)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:103)
	at org.elasticsearch.search.aggregations.AggregationParseElement.parse(AggregationParseElement.java:60)
```

**Comment by cbuescher** [comment]
Looking at DateTimeZone#forID, it seems like Joda couldn't look up the zone from the internal Provider class that Joda loads at startup and holds the time zone data. This seems pluggable, but there are two providers in Joda already, UTCProvider which only knows UTC time zone and ZoneInfoProvider, which should be used by default. This one seems to gets loaded using the class loader from "org/joda/time/tz/data" when the first ISOChronology instance (also internal Joda Singleton) gets initialized.
I will try adding logging to both tests to confirm that if this failure happens again, we see which provider is used.

**Comment by cbuescher** [comment]
Another one: http://build-us-00.elastic.co/job/es_core_2x_debian/530/

**Comment by cbuescher** [comment]
Added Logging confirmed that Joda uses the UTCProvider instead of the ZoneInfoProvider that usually should load the time zone data file:
```
[2015-11-04 09:57:06,003][DEBUG][messy.tests              ] Provider for DateTimeZone [class org.joda.time.tz.UTCProvider] 
  1> [2015-11-04 09:57:06,003][DEBUG][messy.tests              ] Available time zone IDs [[UTC]] 
```
UTCProvider is the last fallback when loading the Provider in `DateTimeZone#getDefaultProvider()`, before that it should succeed in loading resources from "org/joda/time/tz/data" in the joda-jar. So either the jar is broken or it doesn't even find it, which is strange given that its a Joda class trying to load it.
Joda seems pretty good at swallowing exceptions, but it looks like it should write something to Std.err if something goes wrong while loading. Unfortunately haven't found anything in the CI output so far.

**Comment by s1monw** [comment]
yet another one - all on 2.x http://build-us-00.elastic.co/job/es_core_2.x_oel_6/1056/

**Comment by s1monw** [comment]
@cbuescher I think I know why this fails... when you look at the latest failure I pasted above you can see that:

```
Suite: org.elasticsearch.script.groovy.GroovySecurityTests
Completed [1/43] on J0 in 2.87s, 1 test
....
....
....
groovy/target/J0/temp/org.elasticsearch.messy.tests.DateHistogramTests_EFD276EEE9F95729-001
  2> NOTE: test params are: codec=Asserting(Lucene54): {date=PostingsFormat(name=Asserting), d=PostingsFormat(name=Asserting), _field_names=PostingsFormat(name=Asserting), _type=PostingsFormat(name=Asserting), dates=PostingsFormat(name=Asserting), _timestamp=PostingsFormat(name=Asserting), _uid=PostingsFormat(name=Asserting), value=PostingsFormat(name=Asserting), _all=PostingsFormat(name=Asserting)}, docValues:{date=DocValuesFormat(name=Lucene54), d=DocValuesFormat(name=Lucene54), dates=DocValuesFormat(name=Lucene54), value=DocValuesFormat(name=Lucene54), _timestamp=DocValuesFormat(name=Lucene54), _version=DocValuesFormat(name=Lucene54)}, sim=RandomSimilarityProvider(queryNorm=false,coord=crazy): {}, locale=sq_AL, timezone=Africa/Ouagadougou
  2> NOTE: Linux 2.6.32-504.23.4.el6.x86_64 amd64/Oracle Corporation 1.8.0_60 (64-bit)/cpus=8,threads=1,free=275774088,total=501219328
  2> NOTE: All tests run in this JVM: [GroovySecurityTests, TransformOnIndexMapperTests, ContextAndHeaderTransportTests, MinTests, CardinalityTests, GeoDistanceTests, DateHistogramTests]
Completed [20/43] on J0 in 3.19s, 33 tests, 2 errors <<< FAILURES!
```

both run in JVM J0. If you run `GroovySecurityTests` on it's own you can see this in the sys.err:

```
java.io.IOException: Resource not found: "org/joda/time/tz/data/ZoneInfoMap" ClassLoader: sun.misc.Launcher$AppClassLoader@14dad5dc
	at org.joda.time.tz.ZoneInfoProvider.openResource(ZoneInfoProvider.java:210)
	at org.joda.time.tz.ZoneInfoProvider.<init>(ZoneInfoProvider.java:127)
	at org.joda.time.tz.ZoneInfoProvider.<init>(ZoneInfoProvider.java:86)
	at org.joda.time.DateTimeZone.getDefaultProvider(DateTimeZone.java:514)
	at org.joda.time.DateTimeZone.getProvider(DateTimeZone.java:413)
	at org.joda.time.DateTimeZone.forTimeZone(DateTimeZone.java:349)
	at org.joda.time.DateTimeZone.getDefault(DateTimeZone.java:157)
	at org.joda.time.chrono.ISOChronology.getInstance(ISOChronology.java:79)
	at org.joda.time.base.BaseDateTime.<init>(BaseDateTime.java:70)
	at org.joda.time.DateTime.<init>(DateTime.java:171)
	at org.codehaus.groovy.vmplugin.v7.IndyInterface.selectMethod(IndyInterface.java:218)
	at a9d18bbccfaebbc0e2a1f5d7d21d068d4a97c59a.run(a9d18bbccfaebbc0e2a1f5d7d21d068d4a97c59a:1)
	at org.elasticsearch.script.groovy.GroovyScriptEngineService$GroovyScript.run(GroovyScriptEngineService.java:296)
	at org.elasticsearch.script.groovy.GroovySecurityTests.doTest(GroovySecurityTests.java:112)
	at org.elasticsearch.script.groovy.GroovySecurityTests.assertFailure(GroovySecurityTests.java:123)
	at org.elasticsearch.script.groovy.GroovySecurityTests.testEvilGroovyScripts(GroovySecurityTests.java:79)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1660)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:866)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:902)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:916)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:50)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:46)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:49)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:65)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:48)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:367)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:809)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:460)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:875)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:777)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:811)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:822)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:46)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:42)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:54)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:65)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:55)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:367)
	at java.lang.Thread.run(Thread.java:745)
```

it tires to load this without permissions and since it's a static block It will never be re-run. The simple fix would be to load that class before we fire up that test statically, what do you think? maybe @rmuir has an opinion

**Comment by s1monw** [comment]
on 2.x this reproduces the issue:

```
 mvn clean test -Dtests.class=*.GroovySecurityTests\|*.DateHistogramTests -Dtests.jvms=1 -Dtests.seed=7AD385E8819A7321
```

**Comment by cbuescher** [comment]
@simonw I see so when GroovySecurityTests is run before the other it messes up the resource loading? Those DateHistogramTests not using groovy scripting should be moved back to core anyway, I was going to do that but didn't want to cover this test failure by it. Should I move the tests or is this something that should somehow be fixed in the test setupt as well?

**Comment by s1monw** [comment]
IMO we should force load that class during bootstrap AND during EsTestCase initialization

**Comment by s1monw** [comment]
This could work but I'd like to consult @rmuir here first
```diff
diff --git a/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java b/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
index 6a72bb4..620333e 100644
--- a/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
+++ b/core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java
@@ -59,6 +59,11 @@ final class Bootstrap {
     private final CountDownLatch keepAliveLatch = new CountDownLatch(1);
     private final Thread keepAliveThread;
 
+    static {
+        // ensure we load all the timezones in the parent classloader with all permissions
+        org.joda.time.DateTimeZone.getDefault();
+    }
+    
     /** creates a new instance */
     Bootstrap() {
         keepAliveThread = new Thread(new Runnable() {
```

**Comment by s1monw** [comment]
@cbuescher can you just put `org.joda.time.DateTimeZone.getDefault();` into the `GroovySecurityTests` for now to make this stop failing...

**Comment by cbuescher** [comment]
@s1monw before any tests are run?

**Comment by s1monw** [comment]
just put it in a static block for now

**Comment by cbuescher** [comment]
Pushed this to 2.x, although it never failed on master so far your reproduction makes it fail there also. Will put this also on master and check if the problem also exists on 2.1.

**Comment by rmuir** [comment]
this is really a bug in groovy though: its missing an AccessController.doPrivileged block

Can you open an issue with them so we don't have the hack forever?

**Comment by rmuir** [comment]
sorry, joda-time rather.

**Comment by rmuir** [comment]
I think if we switch to java time, which runs in the jdk, we won't have issues like this.

**Comment by s1monw** [comment]
I opened an issue

**Comment by jodastephen** [comment]
I looked into this, but it is not clear to me what permissions you believe Joda-Time should be requesting.

**Comment by jasontedor** [comment]
I opened JodaOrg/joda-time#375. I've tested a build that contains that fix and it resolves the issue reported here, #18017 and #18328 (we can remove the hack added in 5563220fe53aae0c5580e05f09606e04668a9da8).
